lnrpc/routerrpc/router_server.go

```diff
@@ -454,39 +454,155 @@ func (s *Server) GetMissionControlConfig(ctx context.Context,
+			),
 		},
-	}, nil
+	}
+
+	// We only populate fields based on the current estimator.
+	switch v := cfg.Estimator.Config().(type) {
+	case routing.AprioriConfig:
+		resp.Config.Model = MissionControlConfig_APRIORI
+		aCfg := AprioriParameters{
+			HalfLifeSeconds: uint64(v.PenaltyHalfLife.Seconds()),
+			HopProbability:  float32(v.AprioriHopProbability),
+			Weight:          float32(v.AprioriWeight),
+		}
+		resp.Config.EstimatorConfig = &MissionControlConfig_Apriori{
+			Apriori: &aCfg,
+		}
+
+	case routing.BimodalConfig:
+		resp.Config.Model = MissionControlConfig_BIMODAL
+		bCfg := BimodalParameters{
+			NodeWeight: float32(v.BimodalNodeWeight),
+			ScaleMsat:  uint64(v.BimodalScaleMsat),
+			HalfLife:   uint64(v.BimodalHalfLife.Seconds()),
+		}
+		resp.Config.EstimatorConfig = &MissionControlConfig_Bimodal{
+			Bimodal: &bCfg,
+		}
+
+	default:
+		return nil, fmt.Errorf("unknown estimator config type %T", v)
+	}
+
+	return resp, nil
 }
 
-// SetMissionControlConfig returns our current mission control config.
+// SetMissionControlConfig sets parameters in the mission control config.
 func (s *Server) SetMissionControlConfig(ctx context.Context,
 	req *SetMissionControlConfigRequest) (*SetMissionControlConfigResponse,
 	error) {
 
-	cfg := &routing.MissionControlConfig{
-		ProbabilityEstimatorCfg: routing.ProbabilityEstimatorCfg{
-			PenaltyHalfLife: time.Duration(
-				req.Config.HalfLifeSeconds,
-			) * time.Second,
-			AprioriHopProbability: float64(req.Config.HopProbability),
-			AprioriWeight:         float64(req.Config.Weight),
-		},
-		MaxMcHistory: int(req.Config.MaximumPaymentResults),
-		MinFailureRelaxInterval: time.Duration(
+	// Query the current configuration and update it.
+	mcCfg := s.cfg.RouterBackend.MissionControl.GetConfig()
+
+	// We set fields if they were altered.
+	if req.Config.MaximumPaymentResults != 0 {
+		mcCfg.MaxMcHistory = int(
+			req.Config.MaximumPaymentResults,
+		)
+	}
+	if req.Config.MinimumFailureRelaxInterval != 0 {
+		mcCfg.MinFailureRelaxInterval = time.Duration(
 			req.Config.MinimumFailureRelaxInterval,
-		) * time.Second,
+		) * time.Second
+	}
+
+	oldEstimator := mcCfg.Estimator
+	switch req.Config.Model {
+	case MissionControlConfig_APRIORI:
+		// If a switch from one model to another was requested, we
+		// initialize with a default configuration and update the fields
+		// if they were set.
+		aCfg := routing.DefaultAprioriConfig()
+		if v, ok := oldEstimator.(*routing.AprioriEstimator); ok {
+			aCfg = v.AprioriConfig
+		}
+
+		// Check if a configuration has been set.
+		newCfg, ok := req.Config.
+			EstimatorConfig.(*MissionControlConfig_Apriori)
+		if ok {
+			aCfgNew := newCfg.Apriori
+
+			if aCfgNew.HalfLifeSeconds != 0 {
+				aCfg.PenaltyHalfLife = time.Duration(
+					aCfgNew.HalfLifeSeconds,
+				) * time.Second
+			}
+
+			if aCfgNew.HopProbability != 0 {
+				aCfg.AprioriHopProbability = float64(
+					aCfgNew.HopProbability,
+				)
+			}
+
+			if newCfg.Apriori.Weight != 0 {
```

Comment:
This logic looks different (and breaking) from what it was previously. It now seems to use default values that can be overriden, and also weight can't be set to zero here?

Reply:
FYI: in case we need to differentiate the zero value from nil (in case the zero value is a valid one) we usually add an [extra boolean field](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/lightning.proto#L2232)